### PR TITLE
Done

### PR DESCRIPTION
**Pull request by IegorSoboliev**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,53 +1,97 @@
...
+        list[counterSize] = value;
+        counterSize++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndex(index);
+        ensureCapacity();
+        System.arraycopy(list,index, list,index + 1,counterSize - index);
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        System.arraycopy(list, index, list, index + 1, counterSize - index);
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,53 +1,97 @@
...
+        checkIndex(index);
+        list[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Object removed = list[index];
+        System.arraycopy(list, index + 1, list,index,counterSize - index - 1);
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        System.arraycopy(list, index + 1, list, index, counterSize - index - 1);
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,53 +1,97 @@
...
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= counterSize) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+    }
+
+    private void ensureCapacity() {
+        if ((list.length - counterSize) < 1) {
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        if (list.length >= counterSize) {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,53 +1,97 @@
...
+            Object[] interimList = list;
+            list = new Object[counterSize * 3 / 2 + 1];
+            System.arraycopy(interimList,0, list,0,interimList.length);
```
Comment (Tarabarnik) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
            System.arraycopy(interimList,0, list,0,interimList.length);
```
```suggestion
            System.arraycopy(interimList, 0, list, 0,i nterimList.length);
```

File: src/main/java/core/basesyntax/ArrayList.java (hunk 3 above)
Comment (IegorSoboliev) [reply src/main/java/core/basesyntax/ArrayList.java]: You, probably, mean (list.length == counterSize)

**Pull request closed without merging**